// @forestadmin/datasource-mongoose/dist/utils/schema/fields.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mongoose_1 = require("mongoose");
const filter_operators_1 = __importDefault(require("./filter-operators"));
const schema_1 = __importDefault(require("../../mongoose/schema"));
const helpers_1 = require("../helpers");
const version_manager_1 = __importDefault(require("../version-manager"));
/** Generate forest admin schema from mongoose schema */
class FieldsGenerator {
    static buildFieldsSchema(model, stack = [{ prefix: null, asFields: [], asModels: [] }]) {
        const ourSchema = {};
        const childSchema = schema_1.default.fromModel(model).applyStack(stack);
        // Add columns and many to one relations
        for (const [name, field] of Object.entries(childSchema.fields)) {
            if (name !== 'parent') {
                ourSchema[name] = this.buildColumnSchema(field);
                if (field instanceof mongoose_1.SchemaType && field.options.ref)
                    ourSchema[`${name}__manyToOne`] = this.buildManyToOne(field.options.ref, name);
            }
        }
        // When a prefix is used add pk + fk + many to one relation to the parent.
        if (stack.length > 1) {
            const parentPrefix = stack[stack.length - 2].prefix;
            ourSchema._id = this.buildVirtualPrimaryKey();
            ourSchema.parentId = {
                ...this.buildColumnSchema(childSchema.fields.parent._id),
                isPrimaryKey: false,
                validation: [{ operator: 'Present' }],
            };
            ourSchema.parent = this.buildManyToOne((0, helpers_1.escape)(parentPrefix !== null ? `${model.modelName}.${parentPrefix}` : model.modelName), 'parentId');
        }
        return ourSchema;
    }
    /** Fake primary key that will be used when we're flattening collections */
    static buildVirtualPrimaryKey() {
        return {
            type: 'Column',
            columnType: 'String',
            filterOperators: filter_operators_1.default.getSupportedOperators('String'),
            isPrimaryKey: true,
            isReadOnly: true,
            isSortable: true,
        };
    }
    static buildManyToOne(collection, foreignKey) {
        return {
            type: 'ManyToOne',
            foreignCollection: collection,
            foreignKey,
            foreignKeyTarget: '_id',
        };
    }
    /** Build ColumnSchema from CleanSchema */
    static buildColumnSchema(field) {
        const columnType = this.getColumnType(field);
        const schema = {
            columnType,
            filterOperators: filter_operators_1.default.getSupportedOperators(columnType),
            defaultValue: field.options?.default,
            isPrimaryKey: field.path === '_id',
            isReadOnly: !!field.options?.immutable,
            isSortable: !(columnType instanceof Object || columnType === 'Json'),
            type: 'Column',
            validation: field.isRequired ? [{ operator: 'Present' }] : null,
        };
        if (columnType === 'Enum') {
            schema.enumValues = this.getEnumValues(field);
        }
        return schema;
    }
    /** Compute column type from CleanSchema */
    static getColumnType(field) {
        const columnType = this.getColumnTypeRec(field);
        // Enum fields are promoted to enum instead of string _only_ if they are at the root of the
        // record.
        if (columnType === 'String') {
            const enumValues = this.getEnumValues(field);
            if (enumValues && enumValues.every(v => typeof v === 'string')) {
                return 'Enum';
            }
        }
        return columnType;
    }
    /** Build ColumnType from CleanSchema recursively */
    static getColumnTypeRec(field) {
        if (field instanceof mongoose_1.SchemaType) {
            if (['String', 'Number', 'Date', 'Boolean'].includes(field.instance)) {
                return field.instance;
            }
            if ([version_manager_1.default.ObjectIdTypeName, 'Buffer', 'Decimal128'].includes(field.instance)) {
                return 'String';
            }
            return 'Json';
        }
        if (field['[]']) {
            return [this.getColumnTypeRec(field['[]'])];
        }
        return Object.entries(field).reduce((memo, [name, subSchema]) => ({ ...memo, [name]: this.getColumnTypeRec(subSchema) }), {});
    }
    /** Get enum validator from field definition */
    static getEnumValues(field) {
        return field.options?.enum instanceof Array ? field.options.enum : field.options?.enum?.values;
    }
}
exports.default = FieldsGenerator;
