// @forestadmin/datasource-customizer/dist/decorators/hook/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const aggregate_1 = require("./context/aggregate");
const create_1 = require("./context/create");
const delete_1 = require("./context/delete");
const list_1 = require("./context/list");
const update_1 = require("./context/update");
const hook_1 = __importDefault(require("./hook"));
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
class CollectionHookDecorator extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.hooks = {
            List: new hook_1.default(),
            Create: new hook_1.default(),
            Update: new hook_1.default(),
            Delete: new hook_1.default(),
            Aggregate: new hook_1.default(),
        };
    }
    addHook(position, type, handler) {
        this.hooks[type].addHandler(position, handler);
    }
    async create(caller, data) {
        const beforeContext = new create_1.InternalHookBeforeCreateContext(this.childCollection, caller, data);
        await this.hooks.Create.executeBefore(beforeContext);
        const records = await this.childCollection.create(caller, beforeContext.getData());
        const afterContext = new create_1.HookAfterCreateContext(this.childCollection, caller, data, records);
        await this.hooks.Create.executeAfter(afterContext);
        return records;
    }
    async list(caller, filter, projection) {
        const beforeContext = new list_1.InternalHookBeforeListContext(this.childCollection, caller, filter, projection);
        await this.hooks.List.executeBefore(beforeContext);
        const records = await this.childCollection.list(caller, beforeContext.getFilter(), beforeContext.getProjection());
        const afterContext = new list_1.HookAfterListContext(this.childCollection, caller, filter, projection, records);
        await this.hooks.List.executeAfter(afterContext);
        return records;
    }
    async update(caller, filter, patch) {
        const beforeContext = new update_1.InternalHookBeforeUpdateContext(this.childCollection, caller, filter, patch);
        await this.hooks.Update.executeBefore(beforeContext);
        await this.childCollection.update(caller, beforeContext.getFilter(), beforeContext.patch);
        const afterContext = new update_1.HookAfterUpdateContext(this.childCollection, caller, filter, patch);
        await this.hooks.Update.executeAfter(afterContext);
    }
    async delete(caller, filter) {
        const beforeContext = new delete_1.InternalHookBeforeDeleteContext(this.childCollection, caller, filter);
        await this.hooks.Delete.executeBefore(beforeContext);
        await this.childCollection.delete(caller, beforeContext.getFilter());
        const afterContext = new delete_1.HookAfterDeleteContext(this.childCollection, caller, filter);
        await this.hooks.Delete.executeAfter(afterContext);
    }
    async aggregate(caller, filter, aggregation, limit) {
        const beforeContext = new aggregate_1.InternalHookBeforeAggregateContext(this.childCollection, caller, filter, aggregation, limit);
        await this.hooks.Aggregate.executeBefore(beforeContext);
        const aggregationResult = await this.childCollection.aggregate(caller, beforeContext.getFilter(), beforeContext.getAggregation(), beforeContext.limit);
        const afterContext = new aggregate_1.HookAfterAggregateContext(this.childCollection, caller, filter, aggregation, aggregationResult, limit);
        await this.hooks.Aggregate.executeAfter(afterContext);
        return aggregationResult;
    }
}
exports.default = CollectionHookDecorator;
