// @forestadmin/datasource-toolkit/dist/src/validation/condition-tree.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const field_1 = __importDefault(require("./field"));
const rules_1 = require("./rules");
const errors_1 = require("../errors");
const branch_1 = __importDefault(require("../interfaces/query/condition-tree/nodes/branch"));
const leaf_1 = __importDefault(require("../interfaces/query/condition-tree/nodes/leaf"));
const collection_1 = __importDefault(require("../utils/collection"));
class ConditionTreeValidator {
    static validate(conditionTree, collection) {
        if (conditionTree instanceof branch_1.default) {
            ConditionTreeValidator.validateBranch(conditionTree, collection);
        }
        else if (conditionTree instanceof leaf_1.default) {
            ConditionTreeValidator.validateLeaf(conditionTree, collection);
        }
        else {
            throw new errors_1.ValidationError('Unexpected condition tree type');
        }
    }
    static validateBranch(branch, collection) {
        if (!['And', 'Or'].includes(branch.aggregator)) {
            throw new errors_1.ValidationError(`The given aggregator '${branch.aggregator}' ` +
                `is not supported. The supported values are: ['Or', 'And']`);
        }
        if (!Array.isArray(branch.conditions)) {
            throw new errors_1.ValidationError(`The given conditions '${branch.conditions}' were expected to be an array`);
        }
        for (const condition of branch.conditions) {
            ConditionTreeValidator.validate(condition, collection);
        }
    }
    static validateLeaf(leaf, collection) {
        const fieldSchema = collection_1.default.getFieldSchema(collection, leaf.field);
        ConditionTreeValidator.throwIfOperatorNotAllowedWithColumn(leaf, fieldSchema);
        ConditionTreeValidator.throwIfValueNotAllowedWithOperator(leaf, fieldSchema);
        ConditionTreeValidator.throwIfOperatorNotAllowedWithColumnType(leaf, fieldSchema);
        ConditionTreeValidator.throwIfValueNotAllowedWithColumnType(leaf, fieldSchema);
    }
    static throwIfOperatorNotAllowedWithColumn(conditionTree, columnSchema) {
        const operators = columnSchema.filterOperators;
        if (!operators?.has(conditionTree.operator)) {
            throw new errors_1.ValidationError(`The given operator '${conditionTree.operator}' ` +
                `is not supported by the column: '${conditionTree.field}'.\n${operators?.size
                    ? `The allowed types are: [${[...operators]}]`
                    : 'the column is not filterable'}`);
        }
    }
    static throwIfValueNotAllowedWithOperator(conditionTree, columnSchema) {
        const { value, field } = conditionTree;
        const allowedTypes = rules_1.MAP_ALLOWED_TYPES_FOR_OPERATOR_CONDITION_TREE[conditionTree.operator];
        this.validateValues(field, columnSchema, value, allowedTypes);
    }
    static throwIfOperatorNotAllowedWithColumnType(conditionTree, columnSchema) {
        const allowedOperators = rules_1.MAP_ALLOWED_OPERATORS_FOR_COLUMN_TYPE[columnSchema.columnType];
        if (!allowedOperators.includes(conditionTree.operator)) {
            throw new errors_1.ValidationError(`The given operator '${conditionTree.operator}' ` +
                `is not allowed with the columnType schema: '${columnSchema.columnType}'.\n` +
                `The allowed types are: [${allowedOperators}]`);
        }
    }
    static throwIfValueNotAllowedWithColumnType(conditionTree, columnSchema) {
        const { value, field, operator } = conditionTree;
        // exclude some cases where the value is not related to the columnType of the field
        if (operator !== 'ShorterThan' &&
            operator !== 'LongerThan' &&
            operator !== 'AfterXHoursAgo' &&
            operator !== 'BeforeXHoursAgo' &&
            operator !== 'PreviousXDays' &&
            operator !== 'PreviousXDaysToDate') {
            const types = rules_1.MAP_ALLOWED_TYPES_FOR_COLUMN_TYPE[columnSchema.columnType];
            this.validateValues(field, columnSchema, value, types);
        }
    }
    static validateValues(field, columnSchema, value, allowedTypes) {
        if (Array.isArray(value)) {
            value.forEach(itemValue => field_1.default.validateValue(field, columnSchema, itemValue, allowedTypes));
        }
        else {
            field_1.default.validateValue(field, columnSchema, value, allowedTypes);
        }
    }
}
exports.default = ConditionTreeValidator;
