// @forestadmin/forestadmin-client/dist/utils/context-variables-injector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ContextVariablesInjector {
    static isTreeBranch(filter) {
        return 'aggregator' in filter;
    }
    static injectContextInValueCustom(value, replaceFunction) {
        if (typeof value !== 'string') {
            return value;
        }
        let valueWithContextVariablesInjected = value;
        const regex = /{{([^}]+)}}/g;
        let match = regex.exec(value);
        const encounteredVariables = [];
        while (match) {
            const contextVariableKey = match[1];
            if (!encounteredVariables.includes(contextVariableKey)) {
                valueWithContextVariablesInjected = valueWithContextVariablesInjected.replace(new RegExp(`{{${contextVariableKey}}}`, 'g'), replaceFunction(contextVariableKey));
            }
            encounteredVariables.push(contextVariableKey);
            match = regex.exec(value);
        }
        return valueWithContextVariablesInjected;
    }
    static injectContextInValue(value, contextVariables) {
        return this.injectContextInValueCustom(value, contextVariableKey => String(contextVariables.getValue(contextVariableKey)));
    }
    static injectContextInFilter(filter, contextVariables) {
        if (!filter) {
            return null;
        }
        if (ContextVariablesInjector.isTreeBranch(filter)) {
            return {
                ...filter,
                conditions: filter.conditions.map(condition => {
                    return this.injectContextInFilter(condition, contextVariables);
                }),
            };
        }
        return {
            ...filter,
            value: this.injectContextInValue(filter.value, contextVariables),
        };
    }
}
exports.default = ContextVariablesInjector;
