// @forestadmin/forestadmin-client/dist/utils/context-variables.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const USER_VALUE_PREFIX = 'currentUser.';
const USER_VALUE_TAG_PREFIX = 'currentUser.tags.';
const USER_VALUE_TEAM_PREFIX = 'currentUser.team.';
class ContextVariables {
    constructor({ requestContextVariables = {}, team, user, }) {
        this.requestContextVariables = requestContextVariables;
        this.team = team;
        this.user = user;
    }
    getCurrentUserData(contextVariableKey) {
        if (contextVariableKey.startsWith(USER_VALUE_TEAM_PREFIX)) {
            return this.team[contextVariableKey.substring(USER_VALUE_TEAM_PREFIX.length)];
        }
        if (contextVariableKey.startsWith(USER_VALUE_TAG_PREFIX)) {
            return this.user?.tags?.[contextVariableKey.substring(USER_VALUE_TAG_PREFIX.length)];
        }
        return this.user?.[contextVariableKey.substring(USER_VALUE_PREFIX.length)];
    }
    getValue(contextVariableKey) {
        if (contextVariableKey.startsWith(USER_VALUE_PREFIX)) {
            return this.getCurrentUserData(contextVariableKey);
        }
        return this.requestContextVariables[contextVariableKey];
    }
}
exports.default = ContextVariables;
