// @forestadmin/datasource-toolkit/dist/src/interfaces/query/condition-tree/factory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const branch_1 = __importDefault(require("./nodes/branch"));
const leaf_1 = __importDefault(require("./nodes/leaf"));
const record_1 = __importDefault(require("../../../utils/record"));
const schema_1 = __importDefault(require("../../../utils/schema"));
class ConditionTreeFactory {
    static matchRecords(schema, records) {
        const ids = records.map(r => record_1.default.getPrimaryKey(schema, r));
        return ConditionTreeFactory.matchIds(schema, ids);
    }
    static matchIds(schema, ids) {
        const primaryKeyNames = schema_1.default.getPrimaryKeys(schema);
        if (primaryKeyNames.length === 0) {
            throw new Error('Collection must have at least one primary key');
        }
        for (const name of primaryKeyNames) {
            const operators = schema.fields[name].filterOperators;
            if (!operators?.has('Equal') || !operators?.has('In')) {
                throw new Error(`Field '${name}' must support operators: ['Equal', 'In']`);
            }
        }
        return ConditionTreeFactory.matchFields(primaryKeyNames, ids);
    }
    static union(...trees) {
        return ConditionTreeFactory.group('Or', trees);
    }
    static intersect(...trees) {
        const result = ConditionTreeFactory.group('And', trees);
        const isEmptyAnd = result instanceof branch_1.default &&
            result.aggregator === 'And' &&
            result.conditions.length === 0;
        return isEmptyAnd ? null : result;
    }
    static fromPlainObject(json) {
        if (json === null)
            return null;
        if (ConditionTreeFactory.isLeaf(json)) {
            return new leaf_1.default(json.field, json.operator, json.value);
        }
        if (ConditionTreeFactory.isBranch(json)) {
            const branch = new branch_1.default(json.aggregator, json.conditions.map(subTree => ConditionTreeFactory.fromPlainObject(subTree)));
            return branch.conditions.length !== 1 ? branch : branch.conditions[0];
        }
        throw new Error('Failed to instantiate condition tree from json');
    }
    static matchFields(fields, values) {
        if (values.length === 0)
            return ConditionTreeFactory.MatchNone;
        if (fields.length === 1) {
            const fieldValues = values.map(tuple => tuple[0]);
            return fieldValues.length > 1
                ? new leaf_1.default(fields[0], 'In', fieldValues)
                : new leaf_1.default(fields[0], 'Equal', fieldValues[0]);
        }
        const [firstField, ...otherFields] = fields;
        const groups = new Map();
        for (const [firstValue, ...otherValues] of values) {
            if (groups.has(firstValue))
                groups.get(firstValue).push(otherValues);
            else
                groups.set(firstValue, [otherValues]);
        }
        return ConditionTreeFactory.union(...[...groups.entries()].map(([firstValue, subValues]) => ConditionTreeFactory.intersect(ConditionTreeFactory.matchFields([firstField], [[firstValue]]), ConditionTreeFactory.matchFields(otherFields, subValues))));
    }
    static group(aggregator, trees) {
        const conditions = trees
            .filter(Boolean)
            .reduce((currentConditions, tree) => tree instanceof branch_1.default && tree.aggregator === aggregator
            ? [...currentConditions, ...tree.conditions]
            : [...currentConditions, tree], []);
        if (conditions.length === 1) {
            return conditions[0];
        }
        return new branch_1.default(aggregator, conditions);
    }
    static isLeaf(raw) {
        return typeof raw === 'object' && 'field' in raw && 'operator' in raw;
    }
    static isBranch(raw) {
        return typeof raw === 'object' && 'aggregator' in raw && 'conditions' in raw;
    }
}
exports.default = ConditionTreeFactory;
ConditionTreeFactory.MatchNone = new branch_1.default('Or', []);
ConditionTreeFactory.MatchAll = null;
