// @forestadmin/datasource-customizer/dist/decorators/operators-emulate/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const collection_context_1 = __importDefault(require("../../context/collection-context"));
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
class OperatorsEmulateCollectionDecorator extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.fields = new Map();
    }
    emulateFieldOperator(name, operator) {
        this.replaceFieldOperator(name, operator, null);
    }
    replaceFieldOperator(name, operator, replaceBy) {
        // Check that the collection can actually support our rewriting
        const pks = datasource_toolkit_1.SchemaUtils.getPrimaryKeys(this.childCollection.schema);
        pks.forEach(pk => {
            const schema = this.childCollection.schema.fields[pk];
            const operators = schema.filterOperators;
            if (!operators?.has('Equal') || !operators?.has('In')) {
                throw new Error(`Cannot override operators on collection '${this.name}': ` +
                    `the primary key columns must support 'Equal' and 'In' operators`);
            }
        });
        // Check that targeted field is valid
        const field = this.childCollection.schema.fields[name];
        datasource_toolkit_1.FieldValidator.validate(this, name);
        if (!field)
            throw new Error('Cannot replace operator for relation');
        // Mark the field operator as replaced.
        if (!this.fields.has(name))
            this.fields.set(name, new Map());
        this.fields.get(name).set(operator, replaceBy);
        this.markSchemaAsDirty();
    }
    refineSchema(childSchema) {
        const fields = {};
        for (const [name, schema] of Object.entries(childSchema.fields)) {
            if (this.fields.has(name)) {
                const column = schema;
                fields[name] = {
                    ...column,
                    filterOperators: new Set([
                        ...(column.filterOperators ?? []),
                        ...this.fields.get(name).keys(),
                    ]),
                };
            }
            else {
                fields[name] = schema;
            }
        }
        return { ...childSchema, fields };
    }
    async refineFilter(caller, filter) {
        return filter?.override({
            conditionTree: await filter.conditionTree?.replaceLeafsAsync(leaf => this.replaceLeaf(caller, leaf, [])),
        });
    }
    async replaceLeaf(caller, leaf, replacements) {
        // ConditionTree is targeting a field on another collection => recurse.
        if (leaf.field.includes(':')) {
            const [prefix] = leaf.field.split(':');
            const schema = this.schema.fields[prefix];
            const association = this.dataSource.getCollection(schema.foreignCollection);
            const associationLeaf = await leaf
                .unnest()
                .replaceLeafsAsync(subLeaf => association.replaceLeaf(caller, subLeaf, replacements));
            return associationLeaf.nest(prefix);
        }
        return this.fields.get(leaf.field)?.has(leaf.operator)
            ? this.computeEquivalent(caller, leaf, replacements)
            : leaf;
    }
    async computeEquivalent(caller, leaf, replacements) {
        const handler = this.fields.get(leaf.field)?.get(leaf.operator);
        if (handler) {
            const replacementId = `${this.name}.${leaf.field}[${leaf.operator}]`;
            const subReplacements = [...replacements, replacementId];
            if (replacements.includes(replacementId)) {
                throw new Error(`Operator replacement cycle: ${subReplacements.join(' -> ')}`);
            }
            const result = await handler(leaf.value, new collection_context_1.default(this, caller));
            if (result) {
                let equivalent = result instanceof datasource_toolkit_1.ConditionTree ? result : datasource_toolkit_1.ConditionTreeFactory.fromPlainObject(result);
                equivalent = await equivalent.replaceLeafsAsync(subLeaf => this.replaceLeaf(caller, subLeaf, subReplacements));
                datasource_toolkit_1.ConditionTreeValidator.validate(equivalent, this);
                return equivalent;
            }
        }
        // Query all records on the dataSource and emulate the filter.
        return datasource_toolkit_1.ConditionTreeFactory.matchRecords(this.schema, leaf.apply(await this.list(caller, new datasource_toolkit_1.PaginatedFilter({}), leaf.projection.withPks(this)), this, caller.timezone));
    }
}
exports.default = OperatorsEmulateCollectionDecorator;
