// @forestadmin/datasource-customizer/dist/decorators/search/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const uuid_1 = require("uuid");
const collection_context_1 = __importDefault(require("../../context/collection-context"));
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
class SearchCollectionDecorator extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.replacer = null;
    }
    replaceSearch(replacer) {
        this.replacer = replacer;
    }
    refineSchema(subSchema) {
        return { ...subSchema, searchable: true };
    }
    async refineFilter(caller, filter) {
        // Search string is not significant
        if (!filter?.search?.trim().length) {
            return filter.override({ search: null });
        }
        // Implement search ourselves
        if (this.replacer || !this.childCollection.schema.searchable) {
            const ctx = new collection_context_1.default(this, caller);
            let tree = this.defaultReplacer(filter.search, filter.searchExtended);
            if (this.replacer) {
                const plainTree = await this.replacer(filter.search, filter.searchExtended, ctx);
                tree = datasource_toolkit_1.ConditionTreeFactory.fromPlainObject(plainTree);
            }
            // Note that if no fields are searchable with the provided searchString, the conditions
            // array might be empty, which will create a condition returning zero records
            // (this is the desired behavior).
            return filter.override({
                conditionTree: datasource_toolkit_1.ConditionTreeFactory.intersect(filter.conditionTree, tree),
                search: null,
            });
        }
        // Let subcollection deal with the search
        return filter;
    }
    defaultReplacer(search, extended) {
        const searchableFields = SearchCollectionDecorator.getFields(this.childCollection, extended);
        const conditions = searchableFields
            .map(([field, schema]) => SearchCollectionDecorator.buildCondition(field, schema, search))
            .filter(Boolean);
        return datasource_toolkit_1.ConditionTreeFactory.union(...conditions);
    }
    static buildCondition(field, schema, searchString) {
        const { columnType, enumValues, filterOperators } = schema;
        const isNumber = Number(searchString).toString() === searchString;
        const isUuid = (0, uuid_1.validate)(searchString);
        if (columnType === 'Number' && isNumber && filterOperators?.has('Equal')) {
            return new datasource_toolkit_1.ConditionTreeLeaf(field, 'Equal', Number(searchString));
        }
        if (columnType === 'Enum' && filterOperators?.has('Equal')) {
            const searchValue = SearchCollectionDecorator.lenientFind(enumValues, searchString);
            if (searchValue)
                return new datasource_toolkit_1.ConditionTreeLeaf(field, 'Equal', searchValue);
        }
        if (columnType === 'String') {
            const isCaseSensitive = searchString.toLocaleLowerCase() !== searchString.toLocaleUpperCase();
            const supportsIContains = filterOperators?.has('IContains');
            const supportsContains = filterOperators?.has('Contains');
            const supportsEqual = filterOperators?.has('Equal');
            // Perf: don't use case-insensitive operator when the search string is indifferent to case
            let operator;
            if (supportsIContains && (isCaseSensitive || !supportsContains))
                operator = 'IContains';
            else if (supportsContains)
                operator = 'Contains';
            else if (supportsEqual)
                operator = 'Equal';
            if (operator)
                return new datasource_toolkit_1.ConditionTreeLeaf(field, operator, searchString);
        }
        if (columnType === 'Uuid' && isUuid && filterOperators?.has('Equal')) {
            return new datasource_toolkit_1.ConditionTreeLeaf(field, 'Equal', searchString);
        }
        return null;
    }
    static getFields(collection, extended) {
        const fields = [];
        for (const [name, field] of Object.entries(collection.schema.fields)) {
            if (field.type === 'Column')
                fields.push([name, field]);
            if (extended && (field.type === 'ManyToOne' || field.type === 'OneToOne')) {
                const related = collection.dataSource.getCollection(field.foreignCollection);
                for (const [subName, subField] of Object.entries(related.schema.fields))
                    if (subField.type === 'Column')
                        fields.push([`${name}:${subName}`, subField]);
            }
        }
        return fields;
    }
    static lenientFind(haystack, needle) {
        return (haystack?.find(v => v === needle.trim()) ??
            haystack?.find(v => v.toLocaleLowerCase() === needle.toLocaleLowerCase().trim()));
    }
}
exports.default = SearchCollectionDecorator;
