// @forestadmin/datasource-toolkit/dist/src/interfaces/query/aggregation.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const luxon_1 = require("luxon");
const object_hash_1 = __importDefault(require("object-hash"));
const projection_1 = __importDefault(require("./projection"));
const record_1 = __importDefault(require("../../utils/record"));
class Aggregation {
    get projection() {
        const { field, groups } = this;
        const aggregateFields = [field, ...(groups ?? []).map(b => b.field)].filter(Boolean);
        return new projection_1.default(...aggregateFields);
    }
    constructor(components) {
        this.field = components.field;
        this.operation = components.operation;
        this.groups = components.groups;
    }
    apply(records, timezone, limit) {
        const rows = this.formatSummaries(this.createSummaries(records, timezone));
        rows.sort((r1, r2) => {
            if (r1.value === r2.value)
                return 0;
            return r1.value < r2.value ? 1 : -1;
        });
        if (limit && rows.length > limit) {
            rows.length = limit;
        }
        return rows;
    }
    nest(prefix) {
        if (!prefix || prefix.length === 0) {
            return this;
        }
        let nestedField;
        let nestedGroups;
        if (this.field) {
            nestedField = `${prefix}:${this.field}`;
        }
        if (this.groups)
            nestedGroups = this.groups.map(bucket => ({
                field: `${prefix}:${bucket.field}`,
                operation: bucket.operation,
            }));
        return new Aggregation({ field: nestedField, operation: this.operation, groups: nestedGroups });
    }
    replaceFields(handler) {
        const result = new Aggregation(this);
        if (result.field) {
            result.field = handler(result.field);
        }
        result.groups = result.groups?.map(({ field, operation }) => ({
            field: handler(field),
            operation,
        }));
        return result;
    }
    createSummaries(records, timezone) {
        const groupingMap = {};
        for (const record of records) {
            const group = this.createGroup(record, timezone);
            const uniqueKey = (0, object_hash_1.default)(group);
            const summary = groupingMap[uniqueKey] ?? this.createSummary(group);
            this.updateSummaryInPlace(summary, record);
            groupingMap[uniqueKey] = summary;
        }
        return Object.values(groupingMap);
    }
    formatSummaries(summaries) {
        const { operation } = this;
        return operation === 'Avg'
            ? summaries
                .filter(summary => summary.Count)
                .map(summary => ({
                group: summary.group,
                value: summary.Sum / summary.Count,
            }))
            : summaries.map(summary => ({
                group: summary.group,
                value: operation === 'Count' && !this.field ? summary.starCount : summary[operation],
            }));
    }
    createGroup(record, timezone) {
        const group = {};
        for (const { field, operation } of this.groups ?? []) {
            const groupValue = record_1.default.getFieldValue(record, field);
            group[field] = this.applyDateOperation(groupValue, operation, timezone);
        }
        return group;
    }
    createSummary(group) {
        return {
            group,
            starCount: 0,
            Count: 0,
            Sum: 0,
            Min: undefined,
            Max: undefined,
        };
    }
    updateSummaryInPlace(summary, record) {
        summary.starCount += 1; // i.e: count(*)
        if (this.field) {
            const value = record_1.default.getFieldValue(record, this.field);
            if (value !== undefined && value !== null) {
                const { Min: min, Max: max } = summary;
                summary.Count += 1; // i.e: count(column)
                if (min === undefined || value < min)
                    summary.Min = value;
                if (max === undefined || value > max)
                    summary.Max = value;
            }
            if (typeof value === 'number' && !Number.isNaN(value)) {
                summary.Sum += value;
            }
        }
    }
    applyDateOperation(value, operation, timezone) {
        if (operation) {
            const dateTime = luxon_1.DateTime.fromISO(value).setZone(timezone);
            if (operation === 'Year') {
                return dateTime.toFormat('yyyy-01-01');
            }
            if (operation === 'Month') {
                return dateTime.toFormat('yyyy-LL-01');
            }
            if (operation === 'Day') {
                return dateTime.toFormat('yyyy-LL-dd');
            }
            if (operation === 'Week') {
                return dateTime.startOf('week').toFormat('yyyy-LL-dd');
            }
        }
        return value;
    }
}
exports.default = Aggregation;
