// @forestadmin/datasource-mongoose/dist/utils/pipeline/group.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** Transform a forest admin aggregation into mongo pipeline */
class GroupGenerator {
    static group(aggregation) {
        return [
            {
                $group: {
                    _id: this.computeGroups(aggregation.groups),
                    value: this.computeValue(aggregation),
                },
            },
            {
                $project: {
                    _id: 0,
                    value: '$value',
                    group: this.computeGroupsProjection(aggregation.groups),
                },
            },
        ];
    }
    /** Compute aggregation value */
    static computeValue(aggregation) {
        // Handle count(*) case
        if (!aggregation.field)
            return { $sum: 1 };
        // General case
        const field = `$${aggregation.field.replace(/:/g, '.')}`;
        return aggregation.operation === 'Count'
            ? { $sum: { $cond: [{ $ne: [field, null] }, 1, 0] } }
            : {
                [this.AGGREGATION_OPERATION[aggregation.operation]]: field,
            };
    }
    /** Compute _id field for the $group pipeline stage */
    static computeGroups(groups) {
        return (groups ?? []).reduce((memo, group) => {
            let field = `$${group.field.replace(/:/g, '.')}`;
            if (group.operation) {
                if (group.operation === 'Week') {
                    const date = { $dateTrunc: { date: field, startOfWeek: 'Monday', unit: 'week' } };
                    field = { $dateToString: { format: this.GROUP_OPERATION[group.operation], date } };
                }
                else {
                    field = { $dateToString: { format: this.GROUP_OPERATION[group.operation], date: field } };
                }
            }
            return { ...(memo ?? {}), [group.field]: field };
        }, null);
    }
    /** Move fields in _id to the root of the document */
    static computeGroupsProjection(groups) {
        return groups?.length
            ? groups.reduce((memo, group) => ({ ...memo, [group.field]: `$_id.${group.field}` }), {})
            : { $literal: {} };
    }
}
exports.default = GroupGenerator;
GroupGenerator.AGGREGATION_OPERATION = {
    Sum: '$sum',
    Avg: '$avg',
    Count: '$sum',
    Max: '$max',
    Min: '$min',
};
GroupGenerator.GROUP_OPERATION = {
    Year: '%Y-01-01',
    Month: '%Y-%m-01',
    Day: '%Y-%m-%d',
    Week: '%Y-%m-%d',
};
