// @forestadmin/datasource-mongoose/dist/utils/pipeline/reparent.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = __importDefault(require("../../mongoose/schema"));
/**
 * Generate pipeline to query submodels.
 *
 * The operations make rotations in the documents so that the root is changed to the submodel
 * without loosing the parent (which may be needed later on).
 */
class ReparentGenerator {
    static reparent(model, stack) {
        const schema = schema_1.default.fromModel(model);
        return stack.flatMap((step, index) => {
            if (index === 0)
                return this.unflatten(step.asFields);
            const localSchema = schema.getSubSchema(step.prefix);
            const relativePrefix = stack[index - 1].prefix !== null
                ? step.prefix.substring(stack[index - 1].prefix.length + 1)
                : step.prefix;
            return [
                ...(localSchema.isArray
                    ? this.reparentArray(relativePrefix, localSchema.isLeaf)
                    : this.reparentObject(relativePrefix, localSchema.isLeaf)),
                ...this.unflatten(step.asFields),
            ];
        });
    }
    static reparentArray(prefix, inDoc) {
        return [
            { $unwind: { path: `$${prefix}`, includeArrayIndex: 'index' } },
            {
                $replaceRoot: {
                    newRoot: {
                        $mergeObjects: [
                            inDoc ? { content: `$${prefix}` } : `$${prefix}`,
                            {
                                _id: { $concat: [{ $toString: '$_id' }, `.${prefix}.`, { $toString: '$index' }] },
                                parentId: '$_id',
                                parent: '$$ROOT',
                            },
                        ],
                    },
                },
            },
        ];
    }
    static reparentObject(prefix, inDoc) {
        return [
            {
                $replaceRoot: {
                    newRoot: {
                        $mergeObjects: [
                            inDoc ? { content: `$${prefix}` } : `$${prefix}`,
                            {
                                _id: { $concat: [{ $toString: '$_id' }, `.${prefix}`] },
                                parentId: '$_id',
                                parent: '$$ROOT',
                            },
                        ],
                    },
                },
            },
        ];
    }
    static unflatten(asFields) {
        return asFields.length
            ? [
                { $addFields: Object.fromEntries(asFields.map(f => [f.replace(/\./g, '@@@'), `$${f}`])) },
                { $project: Object.fromEntries(asFields.map(f => [f, 0])) },
            ]
            : [];
    }
}
exports.default = ReparentGenerator;
