// @forestadmin/forestadmin-client/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaService = exports.buildApplicationServices = exports.ForestAdminClientWithCache = exports.ChartHandler = exports.ContextVariables = exports.ContextVariablesInjector = exports.CollectionActionEvent = exports.NonSelectSQLQueryError = exports.EmptySQLQueryError = exports.ChainedSQLQueryError = exports.JTWUnableToVerifyError = exports.JTWTokenExpiredError = void 0;
const build_application_services_1 = __importDefault(require("./build-application-services"));
const forest_admin_client_with_cache_1 = __importDefault(require("./forest-admin-client-with-cache"));
const forest_http_api_1 = __importDefault(require("./permissions/forest-http-api"));
var jwt_token_expired_error_1 = require("./permissions/errors/jwt-token-expired-error");
Object.defineProperty(exports, "JTWTokenExpiredError", { enumerable: true, get: function () { return __importDefault(jwt_token_expired_error_1).default; } });
var jwt_unable_to_verify_error_1 = require("./permissions/errors/jwt-unable-to-verify-error");
Object.defineProperty(exports, "JTWUnableToVerifyError", { enumerable: true, get: function () { return __importDefault(jwt_unable_to_verify_error_1).default; } });
var chained_sql_query_error_1 = require("./permissions/errors/chained-sql-query-error");
Object.defineProperty(exports, "ChainedSQLQueryError", { enumerable: true, get: function () { return __importDefault(chained_sql_query_error_1).default; } });
var empty_sql_query_error_1 = require("./permissions/errors/empty-sql-query-error");
Object.defineProperty(exports, "EmptySQLQueryError", { enumerable: true, get: function () { return __importDefault(empty_sql_query_error_1).default; } });
var non_select_sql_query_error_1 = require("./permissions/errors/non-select-sql-query-error");
Object.defineProperty(exports, "NonSelectSQLQueryError", { enumerable: true, get: function () { return __importDefault(non_select_sql_query_error_1).default; } });
// These types are used for the agent-generator package
var types_1 = require("./permissions/types");
Object.defineProperty(exports, "CollectionActionEvent", { enumerable: true, get: function () { return types_1.CollectionActionEvent; } });
function createForestAdminClient(options) {
    const { optionsWithDefaults, permission, renderingPermission, contextVariables, chartHandler, ipWhitelist, schema, auth, } = (0, build_application_services_1.default)(new forest_http_api_1.default(), options);
    return new forest_admin_client_with_cache_1.default(optionsWithDefaults, permission, renderingPermission, contextVariables, chartHandler, ipWhitelist, schema, auth);
}
exports.default = createForestAdminClient;
__exportStar(require("./charts/types"), exports);
__exportStar(require("./schema/types"), exports);
var context_variables_injector_1 = require("./utils/context-variables-injector");
Object.defineProperty(exports, "ContextVariablesInjector", { enumerable: true, get: function () { return __importDefault(context_variables_injector_1).default; } });
var context_variables_1 = require("./utils/context-variables");
Object.defineProperty(exports, "ContextVariables", { enumerable: true, get: function () { return __importDefault(context_variables_1).default; } });
var chart_handler_1 = require("./charts/chart-handler");
Object.defineProperty(exports, "ChartHandler", { enumerable: true, get: function () { return __importDefault(chart_handler_1).default; } });
var forest_admin_client_with_cache_2 = require("./forest-admin-client-with-cache");
Object.defineProperty(exports, "ForestAdminClientWithCache", { enumerable: true, get: function () { return __importDefault(forest_admin_client_with_cache_2).default; } });
var build_application_services_2 = require("./build-application-services");
Object.defineProperty(exports, "buildApplicationServices", { enumerable: true, get: function () { return __importDefault(build_application_services_2).default; } });
// export is necessary for the agent-generator package
var schema_1 = require("./schema");
Object.defineProperty(exports, "SchemaService", { enumerable: true, get: function () { return __importDefault(schema_1).default; } });
