// @forestadmin/datasource-customizer/dist/decorators/publication-field/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
/** This decorator allows hiding fields */
class PublicationFieldCollectionDecorator extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.unpublished = new Set();
    }
    /** Show/hide fields from the schema */
    changeFieldVisibility(name, visible) {
        const field = this.childCollection.schema.fields[name];
        if (!field) {
            throw new Error(`No such field '${name}'`);
        }
        if (datasource_toolkit_1.SchemaUtils.isPrimaryKey(this.childCollection.schema, name)) {
            throw new Error(`Cannot hide primary key`);
        }
        if (!visible)
            this.unpublished.add(name);
        else
            this.unpublished.delete(name);
        this.markSchemaAsDirty();
    }
    async create(caller, data) {
        const records = await super.create(caller, data);
        return records.map(childRecord => {
            const record = {};
            for (const key of Object.keys(childRecord))
                if (!this.unpublished.has(key))
                    record[key] = childRecord[key];
            return record;
        });
    }
    refineSchema(childSchema) {
        const fields = {};
        for (const [name, field] of Object.entries(childSchema.fields)) {
            if (this.isPublished(name)) {
                fields[name] = field;
            }
        }
        return { ...childSchema, fields };
    }
    isPublished(name) {
        const field = this.childCollection.schema.fields[name];
        return (!this.unpublished.has(name) &&
            // Columns have no special requirements
            (field.type === 'Column' ||
                // Many to one, one to one and one to many need the foreign key to be published
                (field.type === 'ManyToOne' && this.isPublished(field.foreignKey)) ||
                ((field.type === 'OneToOne' || field.type === 'OneToMany') &&
                    this.dataSource.getCollection(field.foreignCollection).isPublished(field.originKey)) ||
                // Many to many relations depend on both foreignKey and originKey to be published
                (field.type === 'ManyToMany' &&
                    this.dataSource.getCollection(field.throughCollection).isPublished(field.foreignKey) &&
                    this.dataSource.getCollection(field.throughCollection).isPublished(field.originKey))));
    }
}
exports.default = PublicationFieldCollectionDecorator;
