// @forestadmin/datasource-toolkit/dist/src/interfaces/query/projection/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = __importDefault(require("../../../utils/schema"));
class Projection extends Array {
    get columns() {
        return this.filter(f => !f.includes(':'));
    }
    get relations() {
        return this.reduce((memo, path) => {
            const index = path.indexOf(':');
            if (index !== -1) {
                const field = path.substring(0, index);
                memo[field] = new Projection(...(memo[field] ?? []), path.substring(index + 1));
            }
            return memo;
        }, {});
    }
    equals(other) {
        return this.length === other.length && this.every(field => other.includes(field));
    }
    replace(handler, bind) {
        return this.map(handler, bind).reduce((memo, newPaths) => {
            if (typeof newPaths === 'string')
                return memo.union([newPaths]);
            return memo.union(newPaths);
        }, new Projection());
    }
    union(...otherProjections) {
        const fields = [this, ...otherProjections].reduce((memo, projection) => (projection ? [...memo, ...projection] : memo), []);
        return new Projection(...new Set(fields));
    }
    apply(records) {
        return records.map(record => this.reproject(record));
    }
    withPks(collection) {
        const result = new Projection(...this);
        for (const pk of schema_1.default.getPrimaryKeys(collection.schema)) {
            if (!result.includes(pk))
                result.push(pk);
        }
        for (const [relation, projection] of Object.entries(this.relations)) {
            const schema = collection.schema.fields[relation];
            const association = collection.dataSource.getCollection(schema.foreignCollection);
            const projectionWithPk = projection.withPks(association).nest(relation);
            for (const field of projectionWithPk) {
                if (!result.includes(field))
                    result.push(field);
            }
        }
        return result;
    }
    nest(prefix) {
        return prefix && prefix.length ? this.map(path => `${prefix}:${path}`) : this;
    }
    unnest() {
        const [prefix] = this[0].split(':');
        if (!this.every(path => path.startsWith(prefix))) {
            throw new Error('Cannot unnest projection.');
        }
        return this.map(path => path.substring(prefix.length + 1));
    }
    reproject(record) {
        let result = null;
        if (record) {
            result = {};
            for (const column of this.columns) {
                result[column] = record[column];
            }
            for (const [relation, projection] of Object.entries(this.relations)) {
                result[relation] = projection.reproject(record[relation]);
            }
        }
        return result;
    }
}
exports.default = Projection;
