// @forestadmin/forestadmin-client/dist/permissions/generate-actions-from-permissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generate_action_identifier_1 = require("./generate-action-identifier");
const types_1 = require("./types");
function generateCollectionPermissions(permissions) {
    return Object.entries(permissions).reduce((acc, [collectionId, collectionPermissions]) => {
        const { collection } = collectionPermissions;
        return {
            ...acc,
            [(0, generate_action_identifier_1.generateCollectionActionIdentifier)(types_1.CollectionActionEvent.Browse, collectionId)]: {
                description: collection.browseEnabled,
            },
            [(0, generate_action_identifier_1.generateCollectionActionIdentifier)(types_1.CollectionActionEvent.Read, collectionId)]: {
                description: collection.readEnabled,
            },
            [(0, generate_action_identifier_1.generateCollectionActionIdentifier)(types_1.CollectionActionEvent.Edit, collectionId)]: {
                description: collection.editEnabled,
            },
            [(0, generate_action_identifier_1.generateCollectionActionIdentifier)(types_1.CollectionActionEvent.Add, collectionId)]: {
                description: collection.addEnabled,
            },
            [(0, generate_action_identifier_1.generateCollectionActionIdentifier)(types_1.CollectionActionEvent.Delete, collectionId)]: {
                description: collection.deleteEnabled,
            },
            [(0, generate_action_identifier_1.generateCollectionActionIdentifier)(types_1.CollectionActionEvent.Export, collectionId)]: {
                description: collection.exportEnabled,
            },
        };
    }, {});
}
function generateCustomActionPermission(collectionId, actions) {
    return Object.entries(actions).reduce((acc, [actionName, actionPermissions]) => {
        return {
            ...acc,
            ...{
                [(0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Approve, actionName, collectionId)]: {
                    description: actionPermissions.userApprovalEnabled,
                    conditions: actionPermissions.userApprovalConditions,
                },
                [(0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.SelfApprove, actionName, collectionId)]: {
                    description: actionPermissions.selfApprovalEnabled,
                },
                [(0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Trigger, actionName, collectionId)]: {
                    description: actionPermissions.triggerEnabled,
                    conditions: actionPermissions.triggerConditions,
                },
                [(0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.RequireApproval, actionName, collectionId)]: {
                    description: actionPermissions.approvalRequired,
                    conditions: actionPermissions.approvalRequiredConditions,
                },
            },
        };
    }, {});
}
function generateActionPermissions(permissions) {
    return Object.entries(permissions).reduce((acc, [collectionId, collectionPermissions]) => {
        const { actions } = collectionPermissions;
        return {
            ...acc,
            ...generateCustomActionPermission(collectionId, actions),
        };
    }, {});
}
function generateActionsGloballyAllowed(permissions) {
    return new Set(Object.entries(permissions)
        .filter(([, permission]) => permission.description === true)
        .map(([action]) => action));
}
function generateActionsByRole(permissions) {
    return new Map(Object.entries(permissions)
        .filter(([, permission]) => typeof permission.description !== 'boolean')
        .map(([name, permission]) => [
        name,
        {
            allowedRoles: new Set(permission.description.roles),
            ...(permission.conditions
                ? {
                    conditionsByRole: new Map(permission.conditions?.map(({ roleId, filter }) => [roleId, filter])),
                }
                : {}),
        },
    ]));
}
function generateActionsFromPermissions(environmentPermissions) {
    if (environmentPermissions === true) {
        return {
            isDevelopment: true,
            actionsGloballyAllowed: new Set(),
            actionsByRole: new Map(),
        };
    }
    const remotePermissions = environmentPermissions;
    const allPermissions = {
        ...generateCollectionPermissions(remotePermissions.collections),
        ...generateActionPermissions(remotePermissions.collections),
    };
    return {
        isDevelopment: false,
        actionsGloballyAllowed: generateActionsGloballyAllowed(allPermissions),
        actionsByRole: generateActionsByRole(allPermissions),
    };
}
exports.default = generateActionsFromPermissions;
