// @forestadmin/forestadmin-client/dist/permissions/permission-with-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generate_action_identifier_1 = require("./generate-action-identifier");
const types_1 = require("./types");
class PermissionServiceWithCache {
    constructor(actionPermissionService, renderingPermissionService) {
        this.actionPermissionService = actionPermissionService;
        this.renderingPermissionService = renderingPermissionService;
    }
    async canOnCollection({ userId, collectionName, event, }) {
        const roleId = await this.getRoleIdForUserId(userId);
        return this.actionPermissionService.can(roleId, (0, generate_action_identifier_1.generateCollectionActionIdentifier)(event, collectionName));
    }
    async canExecuteSegmentQuery(params) {
        return this.renderingPermissionService.canExecuteSegmentQuery(params);
    }
    async canTriggerCustomAction({ userId, collectionName, customActionName, }) {
        const roleId = await this.getRoleIdForUserId(userId);
        return this.actionPermissionService.can(roleId, (0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Trigger, customActionName, collectionName));
    }
    async doesTriggerCustomActionRequiresApproval({ userId, collectionName, customActionName, }) {
        // In development everything is allowed (even RequireApproval)
        // but we don't want to require approvals!
        if (await this.actionPermissionService.isDevelopmentPermission()) {
            return false;
        }
        const roleId = await this.getRoleIdForUserId(userId);
        return this.actionPermissionService.can(roleId, (0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.RequireApproval, customActionName, collectionName));
    }
    async canApproveCustomAction({ userId, collectionName, customActionName, requesterId, }) {
        const actionIdentifier = requesterId === userId
            ? (0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.SelfApprove, customActionName, collectionName)
            : (0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Approve, customActionName, collectionName);
        const roleId = await this.getRoleIdForUserId(userId);
        return this.actionPermissionService.can(roleId, actionIdentifier);
    }
    async getConditionalTriggerCondition({ userId, collectionName, customActionName, }) {
        const roleId = await this.getRoleIdForUserId(userId);
        return this.actionPermissionService.getCustomActionCondition(roleId, (0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Trigger, customActionName, collectionName));
    }
    async getConditionalRequiresApprovalCondition({ userId, collectionName, customActionName, }) {
        const roleId = await this.getRoleIdForUserId(userId);
        return this.actionPermissionService.getCustomActionCondition(roleId, (0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.RequireApproval, customActionName, collectionName));
    }
    async getConditionalApproveCondition({ userId, collectionName, customActionName, }) {
        const roleId = await this.getRoleIdForUserId(userId);
        return this.actionPermissionService.getCustomActionCondition(roleId, (0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Approve, customActionName, collectionName));
    }
    async getConditionalApproveConditions({ collectionName, customActionName, }) {
        return this.actionPermissionService.getAllCustomActionConditions((0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Approve, customActionName, collectionName));
    }
    async getRoleIdsAllowedToApproveWithoutConditions({ collectionName, customActionName, }) {
        return this.actionPermissionService.getRoleIdsAllowedToApproveWithoutConditions((0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Approve, customActionName, collectionName));
    }
    async canRequestCustomActionParameters({ userId, collectionName, customActionName, }) {
        const roleId = await this.getRoleIdForUserId(userId);
        return this.actionPermissionService.can(roleId, (0, generate_action_identifier_1.generateCustomActionIdentifier)(types_1.CustomActionEvent.Trigger, customActionName, collectionName));
    }
    async canExecuteChart({ renderingId, userId, chartRequest, }) {
        return this.renderingPermissionService.canExecuteChart({
            renderingId,
            userId,
            chartRequest,
        });
    }
    async getRoleIdForUserId(userId) {
        return (await this.renderingPermissionService.getUser(userId))?.roleId;
    }
}
exports.default = PermissionServiceWithCache;
