// @forestadmin/forestadmin-client/dist/permissions/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomActionEvent = exports.CollectionActionEvent = exports.PermissionLevel = void 0;
var PermissionLevel;
(function (PermissionLevel) {
    PermissionLevel["Admin"] = "admin";
    PermissionLevel["Developer"] = "developer";
    PermissionLevel["Editor"] = "editor";
    PermissionLevel["User"] = "user";
})(PermissionLevel = exports.PermissionLevel || (exports.PermissionLevel = {}));
var CollectionActionEvent;
(function (CollectionActionEvent) {
    CollectionActionEvent["Browse"] = "browse";
    CollectionActionEvent["Export"] = "export";
    CollectionActionEvent["Read"] = "read";
    CollectionActionEvent["Edit"] = "edit";
    CollectionActionEvent["Delete"] = "delete";
    CollectionActionEvent["Add"] = "add";
})(CollectionActionEvent = exports.CollectionActionEvent || (exports.CollectionActionEvent = {}));
var CustomActionEvent;
(function (CustomActionEvent) {
    CustomActionEvent["Trigger"] = "trigger";
    CustomActionEvent["Approve"] = "approve";
    CustomActionEvent["SelfApprove"] = "self-approve";
    CustomActionEvent["RequireApproval"] = "require-approval";
})(CustomActionEvent = exports.CustomActionEvent || (exports.CustomActionEvent = {}));
