// @forestadmin/datasource-customizer/dist/decorators/rename-collection/collection.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const collection_decorator_1 = __importDefault(require("../collection-decorator"));
/**
 * This decorator renames collections.
 * It should be used with RenameCollectionDataSourceDecorator, and not the raw DataSourceDecorator
 */
class RenameCollectionCollectionDecorator extends collection_decorator_1.default {
    constructor() {
        super(...arguments);
        this.substitutedName = null;
    }
    get name() {
        return this.substitutedName ?? this.childCollection.name;
    }
    /** @internal */
    rename(name) {
        this.substitutedName = name;
        // Invalidate all schemas
        for (const collection of this.dataSource.collections) {
            collection.markSchemaAsDirty();
        }
    }
    refineSchema(childSchema) {
        const fields = {};
        for (const [name, oldSchema] of Object.entries(childSchema.fields)) {
            const schema = { ...oldSchema };
            if (schema.type === 'ManyToOne') {
                schema.foreignCollection = this.getNewName(schema.foreignCollection);
            }
            else if (schema.type === 'OneToMany' || schema.type === 'OneToOne') {
                schema.foreignCollection = this.getNewName(schema.foreignCollection);
            }
            else if (schema.type === 'ManyToMany') {
                schema.throughCollection = this.getNewName(schema.throughCollection);
                schema.foreignCollection = this.getNewName(schema.foreignCollection);
            }
            fields[name] = schema;
        }
        return { ...childSchema, fields };
    }
    getNewName(oldName) {
        return this.dataSource.collections.find(c => c.childCollection.name === oldName).name;
    }
}
exports.default = RenameCollectionCollectionDecorator;
