// @forestadmin/forestadmin-client/dist/charts/chart-handler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const chart_detector_1 = __importDefault(require("../utils/chart-detector"));
const context_variables_injector_1 = __importDefault(require("../utils/context-variables-injector"));
class ChartHandlerService {
    constructor(contextVariablesInstantiator) {
        this.contextVariablesInstantiator = contextVariablesInstantiator;
    }
    async getChartWithContextInjected({ userId, renderingId, chartRequest, }) {
        const contextVariables = await this.contextVariablesInstantiator.buildContextVariables({
            userId,
            renderingId,
            requestContextVariables: chartRequest.contextVariables,
        });
        const chart = { ...chartRequest };
        delete chart.contextVariables;
        if (chart_detector_1.default.isFilterableChart(chart)) {
            chart.filter = context_variables_injector_1.default.injectContextInFilter(chart.filter, contextVariables);
        }
        if (chart_detector_1.default.isAggregatedChart(chart)) {
            chart.aggregator = context_variables_injector_1.default.injectContextInValue(chart.aggregator, contextVariables);
        }
        if (chart_detector_1.default.isLineChart(chart)) {
            chart.timeRange = context_variables_injector_1.default.injectContextInValue(chart.timeRange, contextVariables);
        }
        if (chart_detector_1.default.isObjectiveChart(chart)) {
            chart.objective = Number(context_variables_injector_1.default.injectContextInValue(chart.objective, contextVariables));
        }
        if (chart_detector_1.default.isLeaderboardChart(chart)) {
            chart.limit = Number(context_variables_injector_1.default.injectContextInValue(chart.limit, contextVariables));
        }
        return chart;
    }
    async getQueryForChart({ userId, renderingId, chartRequest, }) {
        const contextVariables = await this.contextVariablesInstantiator.buildContextVariables({
            userId,
            renderingId,
            requestContextVariables: chartRequest.contextVariables,
        });
        const contextVariablesUsed = {};
        const replaceContextVariable = (contextVariableName) => {
            const contextVariableRenamed = contextVariableName.replace(/\./g, '_');
            contextVariablesUsed[contextVariableRenamed] = contextVariables.getValue(contextVariableName);
            return `$${contextVariableRenamed}`;
        };
        const query = context_variables_injector_1.default.injectContextInValueCustom(chartRequest.query, replaceContextVariable);
        return {
            query,
            contextVariables: contextVariablesUsed,
        };
    }
}
exports.default = ChartHandlerService;
