// @forestadmin/datasource-customizer/dist/decorators/hook/context/aggregate.js
"use strict";
/* eslint-disable max-classes-per-file */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HookAfterAggregateContext = exports.InternalHookBeforeAggregateContext = exports.HookBeforeAggregateContext = void 0;
const hook_1 = __importDefault(require("./hook"));
class HookBeforeAggregateContext extends hook_1.default {
    constructor(collection, caller, filter, aggregation, limit) {
        super(collection, caller);
        this._filter = filter;
        this._aggregation = aggregation;
        this._limit = limit;
    }
    get filter() {
        return Object.freeze(this._filter);
    }
    get aggregation() {
        return Object.freeze(this._aggregation);
    }
    get limit() {
        return this._limit;
    }
}
exports.HookBeforeAggregateContext = HookBeforeAggregateContext;
class InternalHookBeforeAggregateContext extends HookBeforeAggregateContext {
    getFilter() {
        return this._filter;
    }
    getAggregation() {
        return this._aggregation;
    }
}
exports.InternalHookBeforeAggregateContext = InternalHookBeforeAggregateContext;
class HookAfterAggregateContext extends HookBeforeAggregateContext {
    constructor(collection, caller, filter, aggregation, aggregateResult, limit) {
        super(collection, caller, filter, aggregation, limit);
        this._aggregateResult = aggregateResult;
    }
    get aggregateResult() {
        return Object.freeze(this._aggregateResult);
    }
}
exports.HookAfterAggregateContext = HookAfterAggregateContext;
