// @forestadmin/datasource-customizer/dist/decorators/hook/context/create.js
"use strict";
/* eslint-disable max-classes-per-file */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HookAfterCreateContext = exports.InternalHookBeforeCreateContext = exports.HookBeforeCreateContext = void 0;
const hook_1 = __importDefault(require("./hook"));
class HookBeforeCreateContext extends hook_1.default {
    constructor(collection, caller, data) {
        super(collection, caller);
        this._data = data;
    }
    get data() {
        return Object.freeze(this._data);
    }
}
exports.HookBeforeCreateContext = HookBeforeCreateContext;
class InternalHookBeforeCreateContext extends HookBeforeCreateContext {
    getData() {
        return this._data;
    }
}
exports.InternalHookBeforeCreateContext = InternalHookBeforeCreateContext;
class HookAfterCreateContext extends HookBeforeCreateContext {
    constructor(collection, caller, data, records) {
        super(collection, caller, data);
        this._records = records;
    }
    get records() {
        return Object.freeze(this._records);
    }
}
exports.HookAfterCreateContext = HookAfterCreateContext;
