// @forestadmin/datasource-customizer/dist/decorators/hook/context/list.js
"use strict";
/* eslint-disable max-classes-per-file */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HookAfterListContext = exports.InternalHookBeforeListContext = exports.HookBeforeListContext = void 0;
const hook_1 = __importDefault(require("./hook"));
class HookBeforeListContext extends hook_1.default {
    constructor(collection, caller, filter, projection) {
        super(collection, caller);
        this._filter = filter;
        this._projection = projection;
    }
    get filter() {
        return Object.freeze(this._filter);
    }
    get projection() {
        return Object.freeze(this._projection);
    }
}
exports.HookBeforeListContext = HookBeforeListContext;
class InternalHookBeforeListContext extends HookBeforeListContext {
    getFilter() {
        return this._filter;
    }
    getProjection() {
        return this._projection;
    }
}
exports.InternalHookBeforeListContext = InternalHookBeforeListContext;
class HookAfterListContext extends HookBeforeListContext {
    constructor(collection, caller, filter, projection, records) {
        super(collection, caller, filter, projection);
        this._records = records;
    }
    get records() {
        return Object.freeze(this._records);
    }
}
exports.HookAfterListContext = HookAfterListContext;
