// @forestadmin/datasource-customizer/dist/datasource-customizer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const collection_customizer_1 = __importDefault(require("./collection-customizer"));
const composite_datasource_1 = __importDefault(require("./decorators/composite-datasource"));
const decorators_stack_1 = __importDefault(require("./decorators/decorators-stack"));
const datasource_1 = __importDefault(require("./decorators/publication-collection/datasource"));
const datasource_2 = __importDefault(require("./decorators/rename-collection/datasource"));
const typing_generator_1 = __importDefault(require("./typing-generator"));
/**
 * Allow to create a new Forest Admin agent from scratch.
 * Builds the application by composing and configuring all the collection decorators.
 *
 * Minimal code to add a datasource
 * @example
 * new AgentBuilder(options)
 *  .addDataSource(new SomeDataSource())
 *  .start();
 */
class DataSourceCustomizer {
    /**
     * Retrieve schema of the agent
     */
    get schema() {
        return this.stack.validation.schema;
    }
    /**
     * Get list of customizable collections
     */
    get collections() {
        return this.stack.validation.collections.map(c => this.getCollection(c.name));
    }
    constructor() {
        this.compositeDataSource = new composite_datasource_1.default();
        this.stack = new decorators_stack_1.default(this.compositeDataSource);
    }
    /**
     * Add a datasource
     * @param factory the datasource to add
     * @param options the options
     */
    addDataSource(factory, options) {
        this.stack.queueCustomization(async (logger) => {
            let dataSource = await factory(logger);
            if (options?.include || options?.exclude) {
                const publicationDecorator = new datasource_1.default(dataSource);
                publicationDecorator.keepCollectionsMatching(options.include, options.exclude);
                dataSource = publicationDecorator;
            }
            if (options?.rename) {
                const renamedDecorator = new datasource_2.default(dataSource);
                renamedDecorator.renameCollections(options.rename);
                dataSource = renamedDecorator;
            }
            this.compositeDataSource.addDataSource(dataSource);
        });
        return this;
    }
    /**
     * Create a new API chart
     * @param name name of the chart
     * @param definition definition of the chart
     * @example
     * .addChart('numCustomers', {
     *   type: 'Value',
     *   render: (context, resultBuilder) => {
     *     return resultBuilder.value(123);
     *   }
     * })
     */
    addChart(name, definition) {
        this.stack.queueCustomization(async () => {
            this.stack.chart.addChart(name, definition);
        });
        return this;
    }
    /**
     * Allow to interact with a decorated collection
     * @param name the name of the collection to manipulate
     * @param handle a function that provide a
     *   collection builder on the given collection name
     * @example
     * .customizeCollection('books', books => books.renameField('xx', 'yy'))
     */
    customizeCollection(name, handle) {
        handle(this.getCollection(name));
        return this;
    }
    /**
     * Get given collection by name
     * @param name name of the collection
     */
    getCollection(name) {
        return new collection_customizer_1.default(this, this.stack, name);
    }
    /**
     * Load a plugin across all collections
     * @param plugin instance of the plugin
     * @param options options which need to be passed to the plugin
     * @example
     * import { advancedExport } from '@forestadmin/plugin-advanced-export';
     *
     * dataSourceCustomizer.use(advancedExportPlugin, { format: 'xlsx' });
     */
    use(plugin, options) {
        this.stack.queueCustomization(async () => {
            await plugin(this, null, options);
        });
        return this;
    }
    async getDataSource(logger) {
        await this.stack.applyQueuedCustomizations(logger);
        return this.stack.dataSource;
    }
    getFactory() {
        return async (logger) => this.getDataSource(logger);
    }
    async updateTypesOnFileSystem(typingsPath, typingsMaxDepth) {
        return typing_generator_1.default.updateTypesOnFileSystem(this.stack.hook, typingsPath, typingsMaxDepth);
    }
}
exports.default = DataSourceCustomizer;
