// @forestadmin/datasource-customizer/dist/decorators/chart/result-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const luxon_1 = require("luxon");
class ResultBuilder {
    value(value, previousValue) {
        return { countCurrent: value, countPrevious: previousValue };
    }
    distribution(obj) {
        return Object.entries(obj).map(([key, value]) => ({ key, value }));
    }
    timeBased(timeRange, values) {
        const format = ResultBuilder.formats[timeRange];
        const formatted = {};
        for (const [date, value] of Object.entries(values)) {
            const label = luxon_1.DateTime.fromISO(date).toFormat(format);
            formatted[label] = (formatted[label] ?? 0) + value;
        }
        const dataPoints = [];
        const dates = Object.keys(values).sort((dateA, dateB) => dateA.localeCompare(dateB));
        const first = luxon_1.DateTime.fromISO(dates[0]).startOf(timeRange.toLowerCase());
        const last = luxon_1.DateTime.fromISO(dates[dates.length - 1]);
        for (let current = first; current <= last; current = current.plus({ [timeRange]: 1 })) {
            const label = current.toFormat(format);
            dataPoints.push({ label, values: { value: formatted[label] ?? 0 } });
        }
        return dataPoints;
    }
    percentage(value) {
        return value;
    }
    objective(value, objective) {
        return { value, objective };
    }
    leaderboard(obj) {
        return this.distribution(obj).sort((a, b) => b.value - a.value);
    }
    smart(data) {
        return data;
    }
}
exports.default = ResultBuilder;
ResultBuilder.formats = {
    Day: 'dd/MM/yyyy',
    Week: "'W'W-kkkk",
    Month: 'MMM yy',
    Year: 'yyyy',
};
