// @forestadmin/datasource-toolkit/dist/src/interfaces/query/condition-tree/nodes/branch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = __importDefault(require("./base"));
const projection_1 = __importDefault(require("../../projection"));
class ConditionTreeBranch extends base_1.default {
    get projection() {
        return this.conditions.reduce((memo, condition) => memo.union(condition.projection), new projection_1.default());
    }
    constructor(aggregator, conditions) {
        super();
        this.aggregator = aggregator;
        this.conditions = conditions;
    }
    forEachLeaf(handler) {
        this.conditions.forEach(c => c.forEachLeaf(handler));
    }
    everyLeaf(handler) {
        return this.conditions.every(c => c.everyLeaf(handler));
    }
    someLeaf(handler) {
        return this.conditions.some(c => c.someLeaf(handler));
    }
    inverse() {
        const aggregator = this.aggregator === 'Or' ? 'And' : 'Or';
        return new ConditionTreeBranch(aggregator, this.conditions.map(c => c.inverse()));
    }
    replaceLeafs(handler, bind) {
        return new ConditionTreeBranch(this.aggregator, this.conditions.map(c => c.replaceLeafs(handler, bind)));
    }
    async replaceLeafsAsync(handler, bind) {
        return new ConditionTreeBranch(this.aggregator, await Promise.all(this.conditions.map(c => c.replaceLeafsAsync(handler, bind))));
    }
    match(record, collection, timezone) {
        return this.aggregator === 'And'
            ? this.conditions.every(c => c.match(record, collection, timezone))
            : this.conditions.some(c => c.match(record, collection, timezone));
    }
}
exports.default = ConditionTreeBranch;
