// @forestadmin/datasource-customizer/dist/plugins/import-field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
async function importField(dataSourceCustomizer, collectionCustomizer, options) {
    const name = options.name;
    const { schema } = options.path.split(':').reduce((memo, field) => {
        const collection = dataSourceCustomizer.getCollection(memo.collection);
        const fieldSchema = collection.schema.fields[field];
        if (fieldSchema === undefined) {
            throw new Error(`Field ${field} not found in collection ${collection.name}`);
        }
        if (fieldSchema.type === 'Column')
            return { schema: fieldSchema };
        if (fieldSchema.type === 'ManyToOne' || fieldSchema.type === 'OneToOne')
            return { collection: fieldSchema.foreignCollection };
        throw new Error('Invalid options.path');
    }, { collection: collectionCustomizer.name });
    collectionCustomizer.addField(name, {
        columnType: schema.columnType,
        defaultValue: schema.defaultValue,
        dependencies: [options.path],
        getValues: records => records.map(r => datasource_toolkit_1.RecordUtils.getFieldValue(r, options.path)),
        enumValues: schema.enumValues,
    });
    if (!schema.isReadOnly && !options.readonly) {
        collectionCustomizer.replaceFieldWriting(name, value => {
            const path = options.path.split(':');
            const writingPath = {};
            path.reduce((nestedPath, currentPath, index) => {
                nestedPath[currentPath] = index === path.length - 1 ? value : {};
                return nestedPath[currentPath];
            }, writingPath);
            return writingPath;
        });
    }
    if (schema.isReadOnly && options.readonly === false) {
        throw new Error(`Readonly option should not be false because the field "${options.path}" is not writable`);
    }
    for (const operator of schema.filterOperators) {
        const handler = value => ({ field: options.path, operator, value });
        collectionCustomizer.replaceFieldOperator(name, operator, handler);
    }
    if (schema.isSortable) {
        collectionCustomizer.replaceFieldSorting(name, [{ field: options.path, ascending: true }]);
    }
}
exports.default = importField;
