// @forestadmin/datasource-toolkit/dist/src/interfaces/query/condition-tree/transforms/comparison.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const factory_1 = __importDefault(require("../factory"));
exports.default = () => ({
    Blank: [
        {
            dependsOn: ['In'],
            forTypes: ['String'],
            replacer: leaf => leaf.override({ operator: 'In', value: [null, ''] }),
        },
        {
            dependsOn: ['Missing'],
            replacer: leaf => leaf.override({ operator: 'Missing' }),
        },
    ],
    Missing: [
        {
            dependsOn: ['Equal'],
            replacer: leaf => leaf.override({ operator: 'Equal', value: null }),
        },
    ],
    Present: [
        {
            dependsOn: ['NotIn'],
            forTypes: ['String'],
            replacer: leaf => leaf.override({ operator: 'NotIn', value: [null, ''] }),
        },
        {
            dependsOn: ['NotEqual'],
            replacer: leaf => leaf.override({ operator: 'NotEqual', value: null }),
        },
    ],
    Equal: [
        {
            dependsOn: ['In'],
            replacer: leaf => leaf.override({ operator: 'In', value: [leaf.value] }),
        },
    ],
    In: [
        {
            dependsOn: ['Equal'],
            replacer: leaf => factory_1.default.union(...leaf.value.map(item => leaf.override({ operator: 'Equal', value: item }))),
        },
    ],
    NotEqual: [
        {
            dependsOn: ['NotIn'],
            replacer: leaf => leaf.override({ operator: 'NotIn', value: [leaf.value] }),
        },
    ],
    NotIn: [
        {
            dependsOn: ['NotEqual'],
            replacer: leaf => factory_1.default.intersect(...leaf.value.map(item => leaf.override({ operator: 'NotEqual', value: item }))),
        },
    ],
});
