// @forestadmin/datasource-toolkit/dist/src/interfaces/query/condition-tree/transforms/time.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const luxon_1 = require("luxon");
const factory_1 = __importDefault(require("../factory"));
function format(value) {
    return value.toUTC().toISO({ suppressMilliseconds: true });
}
function compare(operator, dateFn) {
    return {
        dependsOn: [operator],
        forTypes: ['Date', 'Dateonly'],
        replacer: (leaf, tz) => {
            const now = luxon_1.DateTime.utc().setZone(tz);
            return leaf.override({ operator, value: format(dateFn(now, leaf.value)) });
        },
    };
}
function interval(startFn, endFn) {
    return {
        dependsOn: ['LessThan', 'GreaterThan'],
        forTypes: ['Date', 'Dateonly'],
        replacer: (leaf, tz) => {
            const now = luxon_1.DateTime.utc().setZone(tz);
            return factory_1.default.intersect(leaf.override({ operator: 'GreaterThan', value: format(startFn(now, leaf.value)) }), leaf.override({ operator: 'LessThan', value: format(endFn(now, leaf.value)) }));
        },
    };
}
function previousInterval(duration) {
    return interval(now => now.minus({ [duration]: 1 }).startOf(duration), now => now.startOf(duration));
}
function previousIntervalToDate(duration) {
    return interval(now => now.startOf(duration), now => now);
}
exports.default = () => ({
    Before: [compare('LessThan', (_, value) => luxon_1.DateTime.fromISO(value))],
    After: [compare('GreaterThan', (_, value) => luxon_1.DateTime.fromISO(value))],
    Past: [compare('LessThan', now => now)],
    Future: [compare('GreaterThan', now => now)],
    BeforeXHoursAgo: [compare('LessThan', (now, value) => now.minus({ hours: value }))],
    AfterXHoursAgo: [compare('GreaterThan', (now, value) => now.minus({ hours: value }))],
    PreviousWeekToDate: [previousIntervalToDate('week')],
    PreviousMonthToDate: [previousIntervalToDate('month')],
    PreviousQuarterToDate: [previousIntervalToDate('quarter')],
    PreviousYearToDate: [previousIntervalToDate('year')],
    Yesterday: [previousInterval('day')],
    PreviousWeek: [previousInterval('week')],
    PreviousMonth: [previousInterval('month')],
    PreviousQuarter: [previousInterval('quarter')],
    PreviousYear: [previousInterval('year')],
    PreviousXDaysToDate: [
        interval((now, value) => now.minus({ days: value }).startOf('day'), now => now),
    ],
    PreviousXDays: [
        interval((now, value) => now.minus({ days: value }).startOf('day'), now => now.startOf('day')),
    ],
    Today: [
        interval(now => now.startOf('day'), now => now.plus({ day: 1 }).startOf('day')),
    ],
});
