// @forestadmin/forestadmin-client/dist/schema/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const crypto_1 = __importDefault(require("crypto"));
const json_api_serializer_1 = __importDefault(require("json-api-serializer"));
const server_1 = __importDefault(require("../utils/server"));
class SchemaService {
    constructor(options) {
        this.options = options;
    }
    async postSchema(schema) {
        const apimap = SchemaService.serialize(schema);
        const shouldSend = await this.doServerWantsSchema(apimap.meta.schemaFileHash);
        if (shouldSend) {
            await server_1.default.query(this.options, 'post', '/forest/apimaps', {}, apimap);
        }
        return shouldSend;
    }
    static serialize(schema) {
        const data = schema.collections.map(c => ({ id: c.name, ...c }));
        const schemaFileHash = crypto_1.default.createHash('sha1').update(JSON.stringify(schema)).digest('hex');
        return SchemaService.serializer.serialize('collections', data, {
            ...schema.metadata,
            schemaFileHash,
        });
    }
    static get serializer() {
        const serializer = new json_api_serializer_1.default();
        serializer.register('collections', {
            // Pass the metadata provided to the serialization fn
            topLevelMeta: (extraData) => extraData,
            relationships: {
                segments: { type: 'segments' },
                actions: { type: 'actions' },
            },
        });
        serializer.register('segments', {});
        serializer.register('actions', {});
        return serializer;
    }
    async doServerWantsSchema(hash) {
        // Check if the schema was already sent by another agent
        const { sendSchema } = await server_1.default.query(this.options, 'post', '/forest/apimaps/hashcheck', {}, { schemaFileHash: hash });
        return sendSchema;
    }
}
exports.default = SchemaService;
