// @forestadmin/datasource-toolkit/dist/src/interfaces/query/sort/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const record_1 = __importDefault(require("../../../utils/record"));
const projection_1 = __importDefault(require("../projection"));
class Sort extends Array {
    get projection() {
        return new projection_1.default(...this.map(s => s.field));
    }
    replaceClauses(callback) {
        return this.map(ob => callback(ob)).reduce((memo, cbResult) => {
            return Array.isArray(cbResult) ? new Sort(...memo, ...cbResult) : new Sort(...memo, cbResult);
        }, new Sort());
    }
    nest(prefix) {
        return prefix && prefix.length
            ? this.map(ob => ({ field: `${prefix}:${ob.field}`, ascending: ob.ascending }))
            : this;
    }
    inverse() {
        return this.map(({ field, ascending }) => ({ field, ascending: !ascending }));
    }
    unnest() {
        const [prefix] = this[0].field.split(':');
        if (!this.every(ob => ob.field.startsWith(prefix))) {
            throw new Error('Cannot unnest sort.');
        }
        return this.map(({ field, ascending }) => ({
            field: field.substring(prefix.length + 1),
            ascending,
        }));
    }
    apply(records) {
        const { length } = this;
        return records.sort((a, b) => {
            for (let i = 0; i < length; i += 1) {
                const { field, ascending } = this[i];
                const valueOnA = record_1.default.getFieldValue(a, field);
                const valueOnB = record_1.default.getFieldValue(b, field);
                if (valueOnA < valueOnB)
                    return ascending ? -1 : 1;
                if (valueOnA > valueOnB)
                    return ascending ? 1 : -1;
            }
            return 0;
        });
    }
}
exports.default = Sort;
