// @forestadmin/datasource-customizer/dist/decorators/datasource-decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const datasource_toolkit_1 = require("@forestadmin/datasource-toolkit");
class DataSourceDecorator extends datasource_toolkit_1.BaseDataSource {
    get schema() {
        return this.childDataSource.schema;
    }
    constructor(childDataSource, CollectionDecoratorCtor) {
        super();
        this.addCollectionToChildDataSource = childDataSource.addCollection.bind(childDataSource);
        Reflect.defineProperty(childDataSource, 'addCollection', {
            value: this.addCollectionObserver.bind(this),
        });
        this.childDataSource = childDataSource;
        this.CollectionDecoratorCtor = CollectionDecoratorCtor;
        this.childDataSource.collections.forEach(collection => this.addCollection(new this.CollectionDecoratorCtor(collection, this)));
    }
    renderChart(caller, name) {
        return this.childDataSource.renderChart(caller, name);
    }
    addCollectionObserver(collection) {
        this.addCollectionToChildDataSource(collection);
        this.addCollection(new this.CollectionDecoratorCtor(collection, this));
    }
}
exports.default = DataSourceDecorator;
